src/relay/backend/contrib/tensorrt/enable_tensorrt.cc

```diff
@@ -0,0 +1,572 @@
+        {"nn.avg_pool2d", AvgPool2DOpChecker},
+        {"nn.global_max_pool2d", GlobalPool2DOpChecker},
+        {"nn.global_avg_pool2d", GlobalPool2DOpChecker},
+        {"exp", AlwaysChecker},
+        {"log", AlwaysChecker},
+        {"sqrt", AlwaysChecker},
+        {"abs", AlwaysChecker},
+        {"negative", AlwaysChecker},
+        {"nn.batch_flatten", AlwaysChecker},
+        {"expand_dims", ExpandDimsOpChecker},
+        {"squeeze", SqueezeOpChecker},
+        {"concatenate", ConcatenateOpChecker},
+        {"nn.conv2d_transpose", Conv2DTransposeOpChecker},
+        {"transpose", TransposeOpChecker},
+        {"reshape", ReshapeOpChecker},
+        {"nn.pad", PadOpChecker},
+        {"sum", ReduceOpChecker},
+        {"prod", ReduceOpChecker},
+        {"max", ReduceOpChecker},
+        {"min", ReduceOpChecker},
+        {"mean", ReduceOpChecker},
+        {"contrib.adaptive_max_pool2d", AdapativePool2DOpChecker},
+        {"contrib.adaptive_avg_pool2d", AdapativePool2DOpChecker},
+        // Ops which require TRT 5.1.5+
+        {"clip", TrtVersionChecker<5, 1, 5>},
+        {"nn.leaky_relu", TrtVersionChecker<5, 1, 5>},
+        {"sin", TrtVersionChecker<5, 1, 5>},
+        {"cos", TrtVersionChecker<5, 1, 5>},
+        {"atan", TrtVersionChecker<5, 1, 5>},
+        {"ceil", TrtVersionChecker<5, 1, 5>},
+        {"floor", TrtVersionChecker<5, 1, 5>},
+        {"strided_slice", StridedSliceOpChecker},
+        // Ops which require TRT 6.0.1+
+        {"image.resize", ResizeOpChecker}};
+
+// TODO(trevmorr): Use annotation/partitioning framework when that is available.
+/*!
+ * \brief Visitor to check whethere a relay Expr is compatible with a certain
+ * version of TensorRT.
+ */
+class TrtChecker : public ExprVisitor {
+ public:
+  explicit TrtChecker(const std::tuple<int, int, int>& trt_version)
+      : trt_version_(trt_version) {}
+
+  void VisitExpr_(const VarNode* op) {
+    const auto* ttype = op->checked_type().as<TensorTypeNode>();
+    CHECK(ttype);
+    if (!ttype->dtype.is_float()) {
+      compatible_ = false;
+    }
+  }
+
+  void VisitExpr_(const CallNode* call) final {
+    const std::string op_name = (call->op.as<OpNode>())->name;
+    for (size_t i = 0; i < call->args.size(); ++i) {
+      // Workaround for check for transpose of weight.
+      if ((op_name == "nn.conv2d" || op_name == "nn.dense") && i == 1) {
+        auto* transpose = call->args[i].as<CallNode>();
+        if (transpose && transpose->op.as<OpNode>()->name == "transpose") {
+          if (!transpose->args[0].as<VarNode>() &&
+              !call->args[i].as<ConstantNode>()) {
+            compatible_ = false;
+            LOG(INFO) << op_name
+                      << " not supported: must have constant weight.";
+          }
+        } else if (!call->args[i].as<VarNode>() &&
+                   !call->args[i].as<ConstantNode>()) {
+          compatible_ = false;
+          LOG(INFO) << op_name << " not supported: must have constant weight.";
```

Comment:
This seems somewhat complicated. If you are just checking that weights should be ConstantNode, the ConstantNode check should suffice. Also, lets hoist the check outside the for-loop.

Reply:
I agree, this code is to allow a workaround for transposed weights before I added constant-folding. The intention here is to allow the weights to either be:
* VarNode -> (the conv or dense op)
* ConstantNode -> (the conv or dense op)
* VarNode -> Transpose -> (the conv or dense op)
* ConstantNode -> Transpose -> (the conv or dense op)

The order of the transpose is allowed to modify dimension 0 since this transpose is done statically. A transpose anywhere else is not allowed to modify the batch dimension because the TRT API prohibits that. So we need this separate check to catch the transpose before the regular Checker would.

Since I'm using ConstantFolding now, this isn't necessarilly needed. But since it is already implemented it will allow some cases where weights aren't marked as params.